orttraining/orttraining/training_ops/cuda/nn/conv_grad.cc

```diff
@@ -4,114 +4,329 @@
+
+  bool Find(const ConvParams& params, T_Perf* result) {
+    std::lock_guard<OrtMutex> guard(mutex);
+    auto it = map.find(params);
+    if (it == map.end()) {
+      return false;
+    }
+    *result = it->second;
+    return true;
   }
 
-  std::vector<int64_t> strides(conv_attrs_.strides);
-  if (strides.empty()) {
-    strides.resize(rank, 1);
+  void Insert(const ConvParams& params, const T_Perf& algo_perf) {
+    std::lock_guard<OrtMutex> guard(mutex);
+    map[params] = algo_perf;
+  }
+};
+
+AlgoPerfCache<T_BwdDataPerf> bwd_data_algos;
+AlgoPerfCache<T_BwdFilterPerf> bwd_filter_algos;
+
+template <typename T_Perf>
+struct AlgoSearch {};
+
+template <>
+struct AlgoSearch<T_BwdDataPerf> {
+  static constexpr auto DEFAULT_ALGO = CUDNN_CONVOLUTION_BWD_DATA_ALGO_1;
+  static AlgoPerfCache<T_BwdDataPerf>& Cache() { return bwd_data_algos; }
+  static Status FindAlgorithms(const ConvArgs& args, const CUDAExecutionProvider* provider, bool use_more_mem,
+                               std::vector<T_BwdDataPerf>& perf_results) {
+    static const T_BwdDataAlgo algos[] = {
+        CUDNN_CONVOLUTION_BWD_DATA_ALGO_0,        CUDNN_CONVOLUTION_BWD_DATA_ALGO_1,
+        CUDNN_CONVOLUTION_BWD_DATA_ALGO_FFT,      CUDNN_CONVOLUTION_BWD_DATA_ALGO_FFT_TILING,
+        CUDNN_CONVOLUTION_BWD_DATA_ALGO_WINOGRAD, CUDNN_CONVOLUTION_BWD_DATA_ALGO_WINOGRAD_NONFUSED};
+    static constexpr int num_algos = CUDNN_CONVOLUTION_BWD_DATA_ALGO_COUNT;
+    ORT_ENFORCE(sizeof(algos) / sizeof(algos[0]) == num_algos, "Missing cuDNN convolution backward data algorithms.");
+    int perf_count;
+    std::unique_ptr<T_BwdDataPerf[]> candidates(new T_BwdDataPerf[num_algos]);
+    if (args.params.algo_mode == OrtCudnnConvAlgoSearch::HEURISTIC) {
+      CUDNN_RETURN_IF_ERROR(cudnnGetConvolutionBackwardDataAlgorithm_v7(args.handle, args.w_desc, args.y_tensor,
+                                                                        args.conv_desc, args.x_tensor, num_algos,
+                                                                        &perf_count, candidates.get()));
+    } else if (args.params.algo_mode == OrtCudnnConvAlgoSearch::EXHAUSTIVE) {
+      size_t max_workspace_size = use_more_mem ? GetMaxWorkspaceSize(args, algos, num_algos) : AlgoSearchWorkspaceSize;
+      // Use IAllocator's Reserve so the workspace can be freed instead of cached. Because the benchmarking uses a huge
+      // amount of memory, e.g. a few GBs.
+      IAllocatorUniquePtr<void> workspace = provider->GetScratchBuffer<void>(max_workspace_size, true);
+      CUDNN_RETURN_IF_ERROR(cudnnFindConvolutionBackwardDataAlgorithmEx(
+          args.handle, args.w_desc, args.w_data, args.y_tensor, args.dy_data, args.conv_desc, args.x_tensor,
+          args.dx_data, num_algos, &perf_count, candidates.get(), workspace.get(), max_workspace_size));
+    } else {
+      ORT_ENFORCE(false, "Algo mode should be 0, 1 or 2, but got ", args.params.algo_mode);
```

Comment:
OrtCudnnConvAlgoSearch::DEFAULT is missing?

Reply:
ic... OrtCudnnConvAlgoSearch::DEFAULT should been handled in OnlyDefaultAlgorithm(). 

update "Algo mode should be HEURISTIC or EXHAUSTIVE"